user@password JDBC connection strings for application DB no longer work

**Describe the bug**
Startup stalls at "Database setup took ..." when application database JDBC URI uses format like `username:password@host:port` which reads `password@host` as the port, or `username@host:port/db?password=password` which reads `username@host` as the host, but only after migrations has been executed successfully.
Used to work on 0.37.8

**Workaround** Set user and password in the URI parameters - example:
`MB_DB_CONNECTION_URI="mysql://172.17.0.2:3306/metabase?user=myuser&password=mypass"`

**To Reproduce**
1. Start Metabase with something like this (applicable for `postgres://` too) against an empty database:
`MB_DB_CONNECTION_URI="mysql://myuser:mypass@172.17.0.2:3306/metabase"` or
`MB_DB_CONNECTION_URI="mysql://myuser@172.17.0.2:3306/metabase?password=mypass"`
2. It will run through all the migrations, but then it stalls at `Database setup took ...`:
```
...
2021-02-05 12:19:19,224 INFO db.data-migrations :: Running data migration 'migrate-field-types'...
2021-02-05 12:19:19,473 INFO db.data-migrations :: Running data migration 'fix-invalid-field-types'...
2021-02-05 12:19:19,490 INFO db.data-migrations :: Running data migration 'copy-site-url-setting-and-remove-trailing-slashes'...
2021-02-05 12:19:19,503 INFO db.data-migrations :: Running data migration 'ensure-protocol-specified-in-site-url'...
2021-02-05 12:19:19,515 INFO db.data-migrations :: Running data migration 'populate-card-database-id'...
2021-02-05 12:19:19,533 INFO db.data-migrations :: Running data migration 'migrate-humanization-setting'...
2021-02-05 12:19:19,541 INFO db.data-migrations :: Running data migration 'mark-category-fields-as-list'...
2021-02-05 12:19:19,560 INFO db.data-migrations :: Running data migration 'add-legacy-sql-directive-to-bigquery-sql-cards'...
2021-02-05 12:19:19,573 INFO db.data-migrations :: Running data migration 'clear-ldap-user-local-passwords'...
2021-02-05 12:19:19,592 INFO db.data-migrations :: Running data migration 'add-migrated-collections'...
2021-02-05 12:19:19,651 INFO db.data-migrations :: Finished running data migrations.
Database setup took 45.3 s
```
3. After a while it will then fail with the following:
`Caused by: java.sql.SQLException: error parsing url : Incorrect port value : mypass@172.17.0.2` or
`Caused by: java.net.UnknownHostException: myuser@172.17.0.2`

**Information about your Metabase Installation:**
Metabase 0.38.0-rc4

Likely caused by #14275. History #7345. Not exactly related, but similar lack of logging on why it's stalling #10709

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -132,7 +132,7 @@
                       (when (re-find #"postgres(?:ql)?:" conn)
                         ;; jdbc:postgresql: is an opaque URI not subject to further parsing. Strip that off and we can
                         ;; use the structural .getQuery from the URI rather than parsing ourselves
-                        (when-let [details (some-> (.getQuery (URI. (str/replace conn #"jdbc:" "")))
+                        (when-let [details (some-> (.getQuery (URI. (str/replace conn #"^jdbc:" "")))
                                                    codec/form-decode
                                                    walk/keywordize-keys)]
                           (suspicious-postgres-details? details))))

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -22,6 +22,7 @@
   Normally you should use the equivalent functions in `metabase.db.connection` which can be overridden rather than
   using this namespace directly."
   (:require [clojure.java.io :as io]
+            [clojure.string :as str]
             [clojure.tools.logging :as log]
             [clojure.walk :as walk]
             [metabase.config :as config]
@@ -63,6 +64,13 @@
 
 (declare connection-details->jdbc-spec)
 
+(defn- suspicious-postgres-details?
+  "If postgres connection seems iffy. #8908
+  https://github.com/metabase/metabase/issues/8908"
+  [details]
+  (and (= (:ssl details) "true")
+       (not (:sslmode details))))
+
 (defn- parse-connection-string
   "Parse a DB connection URI like
   `postgres://cam@localhost.com:5432/cams_cool_db?ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory` and
@@ -70,42 +78,66 @@
   credentials are included in the like `username:password@host:port`."
   [uri]
   (when-let [[_ _ protocol user pass host port db query] (re-matches jdbc-connection-regex uri)]
-    (connection-details->jdbc-spec
-     (keyword protocol)
-     (u/prog1 (merge {:type (case (keyword protocol)
-                              :postgres   :postgres
-                              :postgresql :postgres
-                              :mysql      :mysql
-                              :h2         :h2)}
+    (let [protocol (case (keyword protocol)
+                     :postgres   :postgres
+                     :postgresql :postgres
+                     :mysql      :mysql
+                     :h2         :h2)
+          details
+          (connection-details->jdbc-spec
+           protocol
+           (merge {:type protocol}
 
-                     (case (keyword protocol)
-                       :h2 {:db db}
-                       {:user     user
-                        :password pass
-                        :host     host
-                        :port     port
-                        :dbname   db})
-                     (some-> query
-                             codec/form-decode
-                             walk/keywordize-keys))
-       ;; If someone is using Postgres and specifies `ssl=true` they might need to specify `sslmode=require`. Let's let
-       ;; them know about that to make their lives a little easier. See https://github.com/metabase/metabase/issues/8908
-       ;; for more details.
-       (when (and (= (:type <>) :postgres)
-                  (= (:ssl <>) "true")
-                  (not (:sslmode <>)))
-         (log/warn (trs "Warning: Postgres connection string with `ssl=true` detected.")
-                   (trs "You may need to add `?sslmode=require` to your application DB connection string.")
-                   (trs "If Metabase fails to launch, please add it and try again.")
-                   (trs "See https://github.com/metabase/metabase/issues/8908 for more details.")))))))
+                  (case (keyword protocol)
+                    :h2 {:db db}
+                    {:user     user
+                     :password pass
+                     :host     host
+                     :port     port
+                     :dbname   db})
+                  (some-> query
+                          codec/form-decode
+                          walk/keywordize-keys)))]
+      ;; If someone is using Postgres and specifies `ssl=true` they might need to specify `sslmode=require`. Let's let
+      ;; them know about that to make their lives a little easier. See
+      ;; https://github.com/metabase/metabase/issues/8908 for more details.
+      {:connection details
+       :diags (cond-> #{:env.warning/inline-credentials}
+                (and (= protocol :postgres) (suspicious-postgres-details? details))
+                (conj :env.warning/postgres-ssl))})))
 
-(defn- ensure-jdbc-protocol
-  "Prepends \"jdbc:\" to the connection-uri string if needed."
+(defn- fixup-connection-string
+  "When we allow a raw connection string as our connection, we still perform a few fixups:
+  - ensure it begins with jdbc:
+  - we allow `postgres:` and must change that to `postgresql:`
+  - warn if postgres ssl settings might cause issues
+
+  Return is a map of {:connection string|spec :diags #{info or warnings}}"
   [connection-uri]
   (when connection-uri
-    (if (re-find #"^jdbc:" connection-uri)
-      connection-uri
-      (str "jdbc:" connection-uri))))
+    (reduce (fn [{:keys [connection] :as m} {:keys [pred diag fix]}]
+              (if (pred connection)
+                (-> m (update :connection fix) (update :diags conj diag))
+                m))
+            {:connection connection-uri
+             :diags #{}}
+            [{:pred #(not (.startsWith ^String % "jdbc:"))
+              :diag :env.info/prepend-jdbc
+              :fix  #(str "jdbc:" %)}
+             {:pred #(re-find #"postgres:" %)
+              :diag :env.info/change-to-postgresql
+              :fix #(str/replace % "postgres:" "postgresql:")}
+             {:pred (fn parse-query-for-postgres
+                      [^String conn]
+                      (when (re-find #"postgres(?:ql)?:" conn)
+                        ;; jdbc:postgresql: is an opaque URI not subject to further parsing. Strip that off and we can
+                        ;; use the structural .getQuery from the URI rather than parsing ourselves
+                        (when-let [details (some-> (.getQuery (URI. (str/replace conn #"jdbc:" "")))
+                                                   codec/form-decode
+                                                   walk/keywordize-keys)]
+                          (suspicious-postgres-details? details))))
+              :fix identity
+              :diag :env.warning/postgres-ssl}])))
 
 (defn old-credential-style?
   "Parse a jdbc connection uri to check for older style credential passing like:
@@ -119,15 +151,35 @@
 (defn- connection-from-jdbc-string
   "If connection string uses the form `username:password@host:port`, use our custom parsing to return a jdbc spec and
   warn about this deprecated behavior. If not, return the jdbc string as is since our parsing does not offer all of
-  the options of using a raw jdbc string."
+  the options of using a raw jdbc string.
+  Return is a map of {:connection string|spec :diags #{info or warnings}}"
   [conn-string]
   (when conn-string
-    (or (when-not (old-credential-style? conn-string)
-          ;; prefer not parsing as we don't handle all features of connection strings
-          (ensure-jdbc-protocol conn-string))
-        (do (log/warn (trs "Warning: using credentials provided inline is deprecated.")
-                      (trs "Change to using the credentials as a query parameter: `?password=your-password&user=user`."))
-            (parse-connection-string conn-string)))))
+    (if (old-credential-style? conn-string)
+        ;; prefer not parsing as we don't handle all features of connection strings
+        (parse-connection-string conn-string)
+        (fixup-connection-string conn-string))))
+
+(defn- log-inline-credentials! []
+  (log/warn
+   (u/format-color 'red
+       (str
+        (trs "Warning: using credentials provided inline is deprecated.")
+        (trs "Change to using the credentials as a query parameter: `?password=your-password&user=user`.")))))
+
+(defn- log-postgres-ssl []
+  (log/warn (trs "Warning: Postgres connection string with `ssl=true` detected.")
+            (trs "You may need to add `?sslmode=require` to your application DB connection string.")
+            (trs "If Metabase fails to launch, please add it and try again.")
+            (trs "See https://github.com/metabase/metabase/issues/8908 for more details.")))
+
+(defn- emit-diags! [diagnostic]
+  (case diagnostic
+    :env.info/change-to-postgresql  (log/info (trs "Replaced 'postgres:' with 'postgresql:' in connection string"))
+    :env.info/prepend-jdbc          (log/info (trs "Prepended 'jdbc:' onto connection string"))
+    :env.warning/inline-credentials (log-inline-credentials!)
+    :env.warning/postgres-ssl       (log-postgres-ssl)
+    (log/warn (trs "Unknown diagnostic in db connection: {0}" diagnostic))))
 
 (def ^:private connection-string-or-spec
   "Uses `:mb-db-connection-uri` and is either:
@@ -135,7 +187,9 @@
   - a db-spec parsed by this code if it does use inline credentials, or
   - nil when this value is not set."
   (delay (when-let [conn-uri (config/config-str :mb-db-connection-uri)]
-           (connection-from-jdbc-string conn-uri))))
+           (let [{:keys [connection diags]} (connection-from-jdbc-string conn-uri)]
+             (run! emit-diags! diags)
+             connection))))
 
 (defn- connection-string-or-spec->db-type [x]
   (cond

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -4,6 +4,10 @@
   enviornment variables e.g. `MB_DB_TYPE`, `MB_DB_HOST`, etc. `MB_DB_CONNECTION_URI` is used preferentially if both
   are specified.
 
+  The `MB_DB_CONNECTION_URI` is unparsed and passed to the jdbc driver with once exception: if it includes the
+  password like `username:password@host:port`. In this case we parse this and log. This functionality will be removed
+  at some point so.
+
   There are two ways we specify JDBC connection information in Metabase code:
 
   1. As a 'connection details' map that is meant to be UI-friendly; this is the actual map we save when creating a
@@ -18,11 +22,15 @@
   Normally you should use the equivalent functions in `metabase.db.connection` which can be overridden rather than
   using this namespace directly."
   (:require [clojure.java.io :as io]
+            [clojure.string :as str]
             [clojure.tools.logging :as log]
+            [clojure.walk :as walk]
             [metabase.config :as config]
             [metabase.db.spec :as db.spec]
             [metabase.util :as u]
-            [metabase.util.i18n :refer [trs]]))
+            [metabase.util.i18n :refer [trs]]
+            [ring.util.codec :as codec])
+  (:import java.net.URI))
 
 (defn- get-db-file
   "Takes a filename and converts it to H2-compatible filename."
@@ -50,20 +58,92 @@
      (let [db-file-name (config/config-str :mb-db-file)]
        (get-db-file db-file-name)))))
 
-(defn- format-connection-uri
+(def ^:private jdbc-connection-regex
+  "Regex to be used only when `:mb-db-connection-uri` includes a password like `username:password@host:port`."
+  #"^(jdbc:)?([^:/@]+)://(?:([^:/@]+)(?::([^:@]+))?@)?([^:@]+)(?::(\d+))?/([^/?]+)(?:\?(.*))?$")
+
+(declare connection-details->jdbc-spec)
+
+(defn- parse-connection-string
+  "Parse a DB connection URI like
+  `postgres://cam@localhost.com:5432/cams_cool_db?ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory` and
+  return a broken-out map. This should not be used unless the connection string uses the old style where the password
+  is included in the like `username:password@host:port`."
+  [uri]
+  (when-let [[_ _ protocol user pass host port db query] (re-matches jdbc-connection-regex uri)]
+    (connection-details->jdbc-spec
+     (keyword protocol)
+     (u/prog1 (merge {:type (case (keyword protocol)
+                              :postgres   :postgres
+                              :postgresql :postgres
+                              :mysql      :mysql
+                              :h2         :h2)}
+
+                     (case (keyword protocol)
+                       :h2 {:db db}
+                       {:user     user
+                        :password pass
+                        :host     host
+                        :port     port
+                        :dbname   db})
+                     (some-> query
+                             codec/form-decode
+                             walk/keywordize-keys))
+       ;; If someone is using Postgres and specifies `ssl=true` they might need to specify `sslmode=require`. Let's let
+       ;; them know about that to make their lives a little easier. See https://github.com/metabase/metabase/issues/8908
+       ;; for more details.
+       (when (and (= (:type <>) :postgres)
+                  (= (:ssl <>) "true")
+                  (not (:sslmode <>)))
+         (log/warn (trs "Warning: Postgres connection string with `ssl=true` detected.")
+                   (trs "You may need to add `?sslmode=require` to your application DB connection string.")
+                   (trs "If Metabase fails to launch, please add it and try again.")
+                   (trs "See https://github.com/metabase/metabase/issues/8908 for more details.")))))))
+
+(defn- ensure-jdbc-protocol
   "Prepends \"jdbc:\" to the connection-uri string if needed."
   [connection-uri]
-  (if-let [uri connection-uri]
-    (if (re-find #"^jdbc:" uri)
-      uri
-      (str "jdbc:" uri))))
+  (when connection-uri
+    (if (re-find #"^jdbc:" connection-uri)
+      connection-uri
+      (str "jdbc:" connection-uri))))
 
-(def ^:private connection-string
-  (delay (format-connection-uri (config/config-str :mb-db-connection-uri))))
+(defn old-password-style?
+  "Parse a jdbc connection uri to check for older style password passing like:
+  mysql://foo:password@172.17.0.2:3306/metabase"
+  [connection-uri]
+  (when connection-uri
+    (let [uri (URI. connection-uri)]
+      ;; this is how clojure.java.jdbc does it
+      (= 2 (count (str/split (.getUserInfo uri) #":"))))))
 
-(defn- connection-string->db-type [s]
-  (when s
-    (let [[_ subprotocol] (re-find #"^(?:jdbc:)?([^:]+):" s)]
+(defn- connection-from-jdbc-string
+  "If connection string uses the form `username:password@host:port`, use our custom parsing to return a jdbc spec and
+  warn about this deprecated behavior. If not, return the jdbc string as is since our parsing does not offer all of
+  the options of using a raw jdbc string."
+  [conn-string]
+  (when conn-string
+    (or (when-not (old-password-style? conn-string)
+          ;; prefer not parsing as we don't handle all features of connection strings
+          (ensure-jdbc-protocol conn-string))
+        (do (log/warn (trs "Warning: using password provided inline is deprecated.")
+                      (trs "Change to using the password as a query parameter: `?password=your-password`."))
+            (parse-connection-string conn-string)))))
+
+(def ^:private connection-string-or-spec
+  "Uses `:mb-db-connection-uri` and is either:
+  - the jdbc connection string if it does not use an inline password, or
+  - a db-spec parsed by this code if it does use an inline password, or
+  - nil when this value is not set."
+  (delay (when-let [conn-uri (config/config-str :mb-db-connection-uri)]
+           (connection-from-jdbc-string conn-uri))))
+
+(defn- connection-string-or-spec->db-type [x]
+  (cond
+    (map? x) (:type x)
+
+    (string? x)
+    (let [[_ subprotocol] (re-find #"^(?:jdbc:)?([^:]+):" x)]
       (try
         (case (keyword subprotocol)
           :postgres   :postgres
@@ -78,7 +158,7 @@
                           {:subprotocol subprotocol})))))))
 
 (def ^:private connection-string-db-type
-  (delay (connection-string->db-type @connection-string)))
+  (delay (connection-string-or-spec->db-type @connection-string-or-spec)))
 
 (def db-type
   "Keyword type name of the application DB details specified by environment variables. Matches corresponding driver
@@ -126,5 +206,5 @@
 (def jdbc-spec
   "`clojure.java.jdbc` spec map for the application DB, using the details map derived from environment variables."
   (delay
-    (or @connection-string
+    (or @connection-string-or-spec
         (connection-details->jdbc-spec @db-type @db-connection-details))))

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -144,7 +144,8 @@
   mysql://foo:password@172.17.0.2:3306/metabase"
   [connection-uri]
   (when connection-uri
-    (let [uri (URI. connection-uri)]
+    ;; strip the jdbc prefix off so its a parseable and not opaque URI
+    (let [uri (URI. (str/replace connection-uri #"^jdbc:" ""))]
       ;; this is how clojure.java.jdbc does it
       (some? (.getUserInfo uri)))))
 

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -22,7 +22,6 @@
   Normally you should use the equivalent functions in `metabase.db.connection` which can be overridden rather than
   using this namespace directly."
   (:require [clojure.java.io :as io]
-            [clojure.string :as str]
             [clojure.tools.logging :as log]
             [clojure.walk :as walk]
             [metabase.config :as config]
@@ -108,14 +107,14 @@
       connection-uri
       (str "jdbc:" connection-uri))))
 
-(defn old-password-style?
+(defn old-credential-style?
   "Parse a jdbc connection uri to check for older style password passing like:
   mysql://foo:password@172.17.0.2:3306/metabase"
   [connection-uri]
   (when connection-uri
     (let [uri (URI. connection-uri)]
       ;; this is how clojure.java.jdbc does it
-      (= 2 (count (str/split (.getUserInfo uri) #":"))))))
+      (some? (.getUserInfo uri)))))
 
 (defn- connection-from-jdbc-string
   "If connection string uses the form `username:password@host:port`, use our custom parsing to return a jdbc spec and
@@ -123,7 +122,7 @@
   the options of using a raw jdbc string."
   [conn-string]
   (when conn-string
-    (or (when-not (old-password-style? conn-string)
+    (or (when-not (old-credential-style? conn-string)
           ;; prefer not parsing as we don't handle all features of connection strings
           (ensure-jdbc-protocol conn-string))
         (do (log/warn (trs "Warning: using password provided inline is deprecated.")

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -4,9 +4,9 @@
   enviornment variables e.g. `MB_DB_TYPE`, `MB_DB_HOST`, etc. `MB_DB_CONNECTION_URI` is used preferentially if both
   are specified.
 
-  The `MB_DB_CONNECTION_URI` is unparsed and passed to the jdbc driver with once exception: if it includes the
-  password like `username:password@host:port`. In this case we parse this and log. This functionality will be removed
-  at some point so.
+  The `MB_DB_CONNECTION_URI` is unparsed and passed to the jdbc driver with once exception: if it includes credentials
+  like `username:password@host:port`. In this case we parse this and log. This functionality will be removed at some
+  point so.
 
   There are two ways we specify JDBC connection information in Metabase code:
 
@@ -58,7 +58,7 @@
        (get-db-file db-file-name)))))
 
 (def ^:private jdbc-connection-regex
-  "Regex to be used only when `:mb-db-connection-uri` includes a password like `username:password@host:port`."
+  "Regex to be used only when `:mb-db-connection-uri` includes credentials like `username:password@host:port`."
   #"^(jdbc:)?([^:/@]+)://(?:([^:/@]+)(?::([^:@]+))?@)?([^:@]+)(?::(\d+))?/([^/?]+)(?:\?(.*))?$")
 
 (declare connection-details->jdbc-spec)
@@ -66,8 +66,8 @@
 (defn- parse-connection-string
   "Parse a DB connection URI like
   `postgres://cam@localhost.com:5432/cams_cool_db?ssl=true&sslfactory=org.postgresql.ssl.NonValidatingFactory` and
-  return a broken-out map. This should not be used unless the connection string uses the old style where the password
-  is included in the like `username:password@host:port`."
+  return a broken-out map. This should not be used unless the connection string uses the old style where the
+  credentials are included in the like `username:password@host:port`."
   [uri]
   (when-let [[_ _ protocol user pass host port db query] (re-matches jdbc-connection-regex uri)]
     (connection-details->jdbc-spec
@@ -108,7 +108,7 @@
       (str "jdbc:" connection-uri))))
 
 (defn old-credential-style?
-  "Parse a jdbc connection uri to check for older style password passing like:
+  "Parse a jdbc connection uri to check for older style credential passing like:
   mysql://foo:password@172.17.0.2:3306/metabase"
   [connection-uri]
   (when connection-uri
@@ -125,14 +125,14 @@
     (or (when-not (old-credential-style? conn-string)
           ;; prefer not parsing as we don't handle all features of connection strings
           (ensure-jdbc-protocol conn-string))
-        (do (log/warn (trs "Warning: using password provided inline is deprecated.")
-                      (trs "Change to using the password as a query parameter: `?password=your-password`."))
+        (do (log/warn (trs "Warning: using credentials provided inline is deprecated.")
+                      (trs "Change to using the credentials as a query parameter: `?password=your-password&user=user`."))
             (parse-connection-string conn-string)))))
 
 (def ^:private connection-string-or-spec
   "Uses `:mb-db-connection-uri` and is either:
-  - the jdbc connection string if it does not use an inline password, or
-  - a db-spec parsed by this code if it does use an inline password, or
+  - the jdbc connection string if it does not use inline credentials, or
+  - a db-spec parsed by this code if it does use inline credentials, or
   - nil when this value is not set."
   (delay (when-let [conn-uri (config/config-str :mb-db-connection-uri)]
            (connection-from-jdbc-string conn-uri))))

--- a/src/metabase/db/env.clj
+++ b/src/metabase/db/env.clj
@@ -164,7 +164,7 @@
   (log/warn
    (u/format-color 'red
        (str
-        (trs "Warning: using credentials provided inline is deprecated.")
+        (trs "Warning: using credentials provided inline is deprecated. ")
         (trs "Change to using the credentials as a query parameter: `?password=your-password&user=user`.")))))
 
 (defn- log-postgres-ssl []

--- a/test/metabase/db/env_test.clj
+++ b/test/metabase/db/env_test.clj
@@ -35,13 +35,20 @@
                 :useUnicode true}]
       (is (= :mysql (#'mdb.env/connection-string-or-spec->db-type spec))))))
 
-(deftest ensure-jdbc-protocol-test
-  (let [conn-uri "postgresql://localhost:metabase?username=johndoe"
-        jdbc-conn-uri (str "jdbc:" conn-uri)]
-    (doseq [[input expected] [[conn-uri jdbc-conn-uri]
-                              [jdbc-conn-uri jdbc-conn-uri]
-                              [nil nil]]]
-      (is (= expected (#'mdb.env/ensure-jdbc-protocol input))))))
+(deftest fixup-connection-string-test
+  (let [correct-uri     "jdbc:postgresql://localhost:metabase?username=johndoe"]
+    (doseq [[input expected-diags expected]
+            [["postgres://localhost:metabase?username=johndoe"
+              #{:env.info/prepend-jdbc :env.info/change-to-postgresql}
+              correct-uri]
+             ["jdbc:postgres://localhost:metabase?username=johndoe"
+              #{:env.info/change-to-postgresql}
+              correct-uri]
+             [correct-uri #{} correct-uri]
+             [nil nil nil]]]
+      (let [{:keys [connection diags]} (#'mdb.env/fixup-connection-string input)]
+        (is (= expected connection) input)
+        (is (= expected-diags diags) input)))))
 
 (deftest old-credential-style?-test
   (are [old? conn-uri] (is (= old? (#'mdb.env/old-credential-style? conn-uri)) conn-uri)
@@ -60,12 +67,40 @@
                 :dbname      "metabase"}]
     (testing "handles mixed (just password in query params) old style just fine"
       (let [conn-uri "mysql://foo@172.17.0.2:3306/metabase?password=password"]
-        (is (= parsed (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+        (is (= {:connection parsed
+                :diags #{:env.warning/inline-credentials}}
+               (#'mdb.env/connection-from-jdbc-string conn-uri)))))
     (testing "When using old-style passwords parses and returns jdbc specs"
       (let [conn-uri "mysql://foo:password@172.17.0.2:3306/metabase"]
-        (is (= parsed (#'mdb.env/connection-from-jdbc-string conn-uri))))))
+        (is (= {:connection parsed
+                :diags #{:env.warning/inline-credentials}}
+               (#'mdb.env/connection-from-jdbc-string conn-uri))))))
   (testing "handles credentials in query params"
     (let [conn-uri "mysql://172.17.0.2:3306/metabase?user=user&password=password"]
-        (is (= (str "jdbc:" conn-uri) (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+      (is (= {:connection (str "jdbc:" conn-uri)
+              :diags #{:env.info/prepend-jdbc}}
+             (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+  (testing "warns about postgres ssl issue #8908"
+    (testing "when it parses due to inline credentials"
+      (let [conn-uri "postgresql://mb@172.17.0.2:5432/metabase?ssl=true"]
+        (is (= {:connection
+                {:ssl "true",
+                 :OpenSourceSubProtocolOverride true,
+                 :password nil,
+                 :type :postgres,
+                 :classname "org.postgresql.Driver",
+                 :subprotocol "postgresql",
+                 :dbname "metabase",
+                 :user "mb",
+                 :subname "//172.17.0.2:5432/metabase"},
+                :diags #{:env.warning/postgres-ssl :env.warning/inline-credentials}}
+               (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+    (testing "when it doesn't parse as well"
+      (doseq [conn-uri ["jdbc:postgresql://172.17.0.2:5432/metabase?user=mb&password=pw&ssl=true"
+                        "postgresql://172.17.0.2:5432/metabase?user=mb&password=pw&ssl=true"
+                        "postgres://172.17.0.2:5432/metabase?user=mb&password=pw&ssl=true"]]
+        (let [{:keys [connection diags]} (#'mdb.env/connection-from-jdbc-string conn-uri)]
+          (is (= "jdbc:postgresql://172.17.0.2:5432/metabase?user=mb&password=pw&ssl=true" connection))
+          (is (contains? diags :env.warning/postgres-ssl))))))
   (testing "handles nil"
     (is (nil? (#'mdb.env/connection-from-jdbc-string nil)))))

--- a/test/metabase/db/env_test.clj
+++ b/test/metabase/db/env_test.clj
@@ -36,7 +36,7 @@
       (is (= :mysql (#'mdb.env/connection-string-or-spec->db-type spec))))))
 
 (deftest fixup-connection-string-test
-  (let [correct-uri     "jdbc:postgresql://localhost:metabase?username=johndoe"]
+  (let [correct-uri "jdbc:postgresql://localhost:metabase?username=johndoe"]
     (doseq [[input expected-diags expected]
             [["postgres://localhost:metabase?username=johndoe"
               #{:env.info/prepend-jdbc :env.info/change-to-postgresql}
@@ -54,6 +54,8 @@
   (are [old? conn-uri] (is (= old? (#'mdb.env/old-credential-style? conn-uri)) conn-uri)
     false "jdbc:mysql://172.17.0.2:3306/metabase?password=password&user=user"
     false "mysql://172.17.0.2:3306/metabase?password=password&user=user"
+    ;; prefixed with jdbc makes an "opaque" URI which doesn't parse
+    true  "jdbc:mysql://foo@172.17.0.2:3306/metabase?password=password"
     true  "mysql://foo@172.17.0.2:3306/metabase?password=password"
     true  "mysql://foo:password@172.17.0.2:3306/metabase"))
 

--- a/test/metabase/db/env_test.clj
+++ b/test/metabase/db/env_test.clj
@@ -43,25 +43,29 @@
                               [nil nil]]]
       (is (= expected (#'mdb.env/ensure-jdbc-protocol input))))))
 
-(deftest old-password-style?-test
-  (are [old? conn-uri] (is (= old? (#'mdb.env/old-password-style? conn-uri)))
-    false  "mysql://foo@172.17.0.2:3306/metabase?password=password"
-    true "mysql://foo:password@172.17.0.2:3306/metabase"))
+(deftest old-credential-style?-test
+  (are [old? conn-uri] (is (= old? (#'mdb.env/old-credential-style? conn-uri)) conn-uri)
+    false "jdbc:mysql://172.17.0.2:3306/metabase?password=password&user=user"
+    false "mysql://172.17.0.2:3306/metabase?password=password&user=user"
+    true  "mysql://foo@172.17.0.2:3306/metabase?password=password"
+    true  "mysql://foo:password@172.17.0.2:3306/metabase"))
 
 (deftest connection-from-jdbc-string-test
-  (testing "handles old style just fine"
-    (let [conn-uri "mysql://foo@172.17.0.2:3306/metabase?password=password"]
-      (is (= (str "jdbc:" conn-uri)
-             (#'mdb.env/connection-from-jdbc-string conn-uri)))))
-  (testing "When using old-style passwords parses and returns jdbc specs"
-    (let [conn-uri "mysql://foo:password@172.17.0.2:3306/metabase"]
-      (is (= {:classname   "org.mariadb.jdbc.Driver"
-              :subprotocol "mysql"
-              :subname     "//172.17.0.2:3306/metabase"
-              :type        :mysql
-              :user        "foo"
-              :password    "password"
-              :dbname      "metabase"}
-             (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+  (let [parsed {:classname   "org.mariadb.jdbc.Driver"
+                :subprotocol "mysql"
+                :subname     "//172.17.0.2:3306/metabase"
+                :type        :mysql
+                :user        "foo"
+                :password    "password"
+                :dbname      "metabase"}]
+    (testing "handles mixed (just password in query params) old style just fine"
+      (let [conn-uri "mysql://foo@172.17.0.2:3306/metabase?password=password"]
+        (is (= parsed (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+    (testing "When using old-style passwords parses and returns jdbc specs"
+      (let [conn-uri "mysql://foo:password@172.17.0.2:3306/metabase"]
+        (is (= parsed (#'mdb.env/connection-from-jdbc-string conn-uri))))))
+  (testing "handles credentials in query params"
+    (let [conn-uri "mysql://172.17.0.2:3306/metabase?user=user&password=password"]
+        (is (= (str "jdbc:" conn-uri) (#'mdb.env/connection-from-jdbc-string conn-uri)))))
   (testing "handles nil"
     (is (nil? (#'mdb.env/connection-from-jdbc-string nil)))))

--- a/test/metabase/db/env_test.clj
+++ b/test/metabase/db/env_test.clj
@@ -2,7 +2,7 @@
   (:require [clojure.test :refer :all]
             [metabase.db.env :as mdb.env]))
 
-(deftest connection-string->db-type-test
+(deftest connection-string-or-spec->db-type-test
   (doseq [[subprotocol expected] {"postgres"   :postgres
                                   "postgresql" :postgres
                                   "mysql"      :mysql
@@ -12,19 +12,56 @@
                                   (str protocol ":abc")
                                   (str protocol ":cam@localhost/my_db?password=123456")
                                   (str protocol "://localhost/my_db")]]
-    (testing (pr-str (list 'connection-string->db-type url))
+    (testing (pr-str (list 'connection-string-or-spec->db-type url))
       (is (= expected
-             (#'mdb.env/connection-string->db-type url)))))
+             (#'mdb.env/connection-string-or-spec->db-type url)))))
   (testing "Should throw an Exception for an unsupported subprotocol"
     (is (thrown-with-msg?
          clojure.lang.ExceptionInfo
          #"Unsupported application database type: \"sqlserver\""
-         (#'mdb.env/connection-string->db-type "jdbc:sqlserver://bad")))))
+         (#'mdb.env/connection-string-or-spec->db-type "jdbc:sqlserver://bad"))))
+  (testing "Finds type from jdbc-spec"
+    (let [spec {:password "password",
+                :characterSetResults "UTF8",
+                :characterEncoding "UTF8",
+                :type :mysql,
+                :classname "org.mariadb.jdbc.Driver",
+                :subprotocol "mysql",
+                :useSSL false,
+                :zeroDateTimeBehavior "convertToNull",
+                :user "foo",
+                :subname "//172.17.0.2:3306/metabase",
+                :useCompression true,
+                :useUnicode true}]
+      (is (= :mysql (#'mdb.env/connection-string-or-spec->db-type spec))))))
 
-(deftest format-connection-uri-test
+(deftest ensure-jdbc-protocol-test
   (let [conn-uri "postgresql://localhost:metabase?username=johndoe"
         jdbc-conn-uri (str "jdbc:" conn-uri)]
     (doseq [[input expected] [[conn-uri jdbc-conn-uri]
                               [jdbc-conn-uri jdbc-conn-uri]
                               [nil nil]]]
-      (is (= expected (#'mdb.env/format-connection-uri input))))))
+      (is (= expected (#'mdb.env/ensure-jdbc-protocol input))))))
+
+(deftest old-password-style?-test
+  (are [old? conn-uri] (is (= old? (#'mdb.env/old-password-style? conn-uri)))
+    false  "mysql://foo@172.17.0.2:3306/metabase?password=password"
+    true "mysql://foo:password@172.17.0.2:3306/metabase"))
+
+(deftest connection-from-jdbc-string-test
+  (testing "handles old style just fine"
+    (let [conn-uri "mysql://foo@172.17.0.2:3306/metabase?password=password"]
+      (is (= (str "jdbc:" conn-uri)
+             (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+  (testing "When using old-style passwords parses and returns jdbc specs"
+    (let [conn-uri "mysql://foo:password@172.17.0.2:3306/metabase"]
+      (is (= {:classname   "org.mariadb.jdbc.Driver"
+              :subprotocol "mysql"
+              :subname     "//172.17.0.2:3306/metabase"
+              :type        :mysql
+              :user        "foo"
+              :password    "password"
+              :dbname      "metabase"}
+             (#'mdb.env/connection-from-jdbc-string conn-uri)))))
+  (testing "handles nil"
+    (is (nil? (#'mdb.env/connection-from-jdbc-string nil)))))

